### PR TITLE
[ENH] Implement full coordinate-set empirical null method 

### LINKED ISSUE #389: Replace existing "empirical" generation of uncorrected p-values with full-blown coordinate shuffling
For ALE and MKDA, we now have both analytical and permutation-based procedures for assigning uncorrected p-values to voxels. However, both methods are conservative, in that they assume voxels are independent, which is not true in practice. I.e., the current empirical permutation approach (which asymptotically converges with the analytical approach) randomly samples a single voxel from each study repeatedly, and builds up a null distribution by repeating the procedure many times. This overestimates the number of truly independent observations, and hence produces conservative correction thresholds (or uncorrected p-values).

We already support p-value generation based on actual shuffling of coordinates in the `FWECorrector`, so the proposal here is really just to provide the option to extract uncorrected p-values that way too. And since the current "empirical" approach is basically inferior to the analytical approach in all respects, we can probably just drop it entirely in favor of the full-blown coordinate shuffling. (I would probably keep these methods around but undocumented; for one thing, they're very helpful in ensuring that the analytical procedures are working as expected, as we can easily test for convergence.)

Probably we should rename the options too, as "empirical" is a bit confusing, IMO. I would suggest calling them "approximate" and "montecarlo", which would intuitively convey to users both that (a) the "approximate" approach doesn't really give the 'correct' answer (because it's biased by the independence assumption), and (b) the "montecarlo" approach is based on resampling and is likely to take a long time. Then the docs could more explicitly lay out the tradeoff, taking pain to emphasize that the "approximate" approach is actually conservative, so if you're willing to accept reduced sensitivity, it makes sense to just always use that.

[EDIT: One other minor issue: our current coordinate shuffling procedures don't take into account spatial relationships *within* a study. I.e., if two peaks overlap in the observed data, the probability of their overlap in any permutation is no greater than two peaks selected at random. I gather the reference implementation of MKDA accounts for the spatial structure in the resampling process, though I'm not sure exactly how that's done. We may want to explore this and do the same thing in NiMARE at some point.]

### PR DESCRIPTION
**Pull request by tsalo**
Closes #389.

Changes proposed in this pull request:

- Add `_determine_histogram_bins()` method to CBMA Estimators. This method determines the bins for the analytic null and the full empirical null, but not the basic (single-coordinate) empirical null, which doesn't use a histogram.
- Add `_run_montecarlo_permutation()` and `_compute_null_reduced_empirical()` methods to base CBMAEstimator.
- Edit `_compute_null_empirical()` to run full coordinate-set method.
- Add `vfwe_only` parameter to the standard Monte Carlo permutation method for FWE correction.
- Directly infer the weighting factor for the kernel in the KDA estimator instead of trying to guess it from MA values.
    - This is currently incompatible with continuous kernels, but I may change that before merging this PR.

### CODE REVIEW

**Comment by tsalo** [comment]
The basic gist here seems to be that we'll need to be careful with our tests since this new null method seems to get killed very easily.

File: nimare/meta/cbma/mkda.py [hunk 1]
```diff
@@ -45,12 +45,18 @@ class MKDADensity(CBMAEstimator):
     """
 
     def __init__(
-        self, kernel_transformer=MKDAKernel, null_method="empirical", n_iters=10000, **kwargs
+        self,
+        kernel_transformer=MKDAKernel,
+        null_method="empirical",
```
Comment (tsalo) [inline nimare/meta/cbma/mkda.py:50]:
```suggestion
        null_method="analytic",
```

File: nimare/meta/cbma/mkda.py [hunk 2]
```diff
@@ -467,7 +494,12 @@ class KDA(CBMAEstimator):
     """
 
     def __init__(
-        self, kernel_transformer=KDAKernel, null_method="empirical", n_iters=10000, **kwargs
+        self,
+        kernel_transformer=KDAKernel,
+        null_method="empirical",
```
Comment (tsalo) [inline nimare/meta/cbma/mkda.py:499]:
```suggestion
        null_method="analytic",
```

**Comment by tsalo** [comment]
`build_docs` is failing on the combination of ALEKernel and KDA estimator because of the way that the KDA analytic null method is implemented. The KDA method uses the smallest observed MA value to define its histogram bins, which is great for the KDA kernel where we don't know the kernel's value, but which is very bad for the ALE kernel, which is continuous, since it results in an oversized histogram.

I still need to dig into the other workflows, which seem to be killed on different tests.

**Comment by tsalo** [comment]
Current failures:

- test_meta_fit_performance[normal_data-ale+analytic-mkda_kernel]
- test_meta_fit_performance[normal_data-ale+analytic-kda_kernel]
- test_meta_fit_performance[normal_data-mkda+empirical-ale_kernel]
- test_corr_transform_performance[fwe_bonferroni-normal_data-ale+analytic-kda_kernel]
- test_corr_transform_performance[fwe_bonferroni-normal_data-ale+analytic-mkda_kernel]
- test_corr_transform_performance[fwe_bonferroni-normal_data-mkda+empirical-ale_kernel]
- test_corr_transform_performance[fdr_indep-normal_data-ale+analytic-kda_kernel]
- test_corr_transform_performance[fdr_indep-normal_data-ale+analytic-mkda_kernel]
- test_corr_transform_performance[fdr_indep-normal_data-mkda+empirical-ale_kernel]
- test_corr_transform_performance[fdr_negcorr-normal_data-ale+analytic-kda_kernel]
- test_corr_transform_performance[fdr_negcorr-normal_data-ale+analytic-mkda_kernel]
- test_corr_transform_performance[fdr_negcorr-normal_data-mkda+empirical-ale_kernel]

**Comment by jdkent** [comment]
It looks like the tests are failing because when I was trying to tease apart whether the estimator/kernel/null_method/corrector combinations "worked", I bundled together my idea of "sensitivity" and "specificity", namely, whether the combination found significant results in the regions where I expected significant results (sensitivity) and non-significant results everywhere else (specificity). Until now, if a combination had poor sensitivity it also had poor specificity(?), but with this pull request, the combinations have either poor sensitivity or specificity. There is a peculiar divergence between the analytic and "empirical" null methods with the ALE estimator with the KDA or MKDA kernels. The analytic null method leads to many voxels being declared significant, but the empirical nulls have the opposite problem. I could not find a problem with how the null were calculated, so I am inclined to think the diverging issues are an idiosyncrasy of using the ALE/(M)KDA Kernel combinations with different null methods.

I'll submit a pull request to this branch fixing the tests once they pass locally.

  | ALEKernel | KDAKernel | MKDAKernel |  
-- | -- | -- | -- | --
ALE | Good | Poor Specificity | Poor Specificity | analytic
  | Good | Poor Sensitivity | Poor Sensitivity | empirical
  | Good | Poor Sensitivity | Poor Sensitivity | reduced empirical
KDA | Good | Good | Good | analytic
  | Good | Good | Good | empirical
  | Good | Good | Good | reduced empirical
MKDA | Poor Sensitivity | Good | Good | analytic
  | Poor Sensitivity | Good | Good | empirical
  | Good | Good | Good | reduced empirical

File: nimare/meta/cbma/base.py [hunk 3]
```diff
@@ -218,18 +204,31 @@ def _summarystat_to_p(self, stat_values, null_method="analytic"):
...
+            assert "values_corr-none_method-reducedEmpirical" in self.null_distributions_.keys()
             p_values = null_to_p(
-                stat_values, self.null_distributions_["empirical_null"], tail="upper"
+                stat_values,
+                self.null_distributions_["values_corr-none_method-reducedEmpirical"],
+                tail="upper",
             )
+
         else:
             raise ValueError("Argument 'null_method' must be one of: 'analytic', 'empirical'.")
```
Comment (jdkent) [inline nimare/meta/cbma/base.py:233]: should we add `reduced_empirical`?
```suggestion
            raise ValueError("Argument 'null_method' must be one of: 'analytic', 'empirical', 'reduced_empirical'.")
```

Comment (tsalo) [reply nimare/meta/cbma/base.py:233]: Not sure. I think @tyarkoni wanted to leave it undocumented, since the reduced empirical method should only be used for testing once we have the full empirical method merged in.

**Comment by jdkent** [comment]
Oh, I have a pull request that _should_ fix the tests tsalo/NiMARE#7, if failed tests are holding this pull request back
